posts/frangeris/01.reusable-dynamic-modal-on-vue3.md

```diff
@@ -0,0 +1,215 @@
+          @click="modal.close()"
+          >✕</label
+        >
+
+        <!-- dynamic components, using model to share values payload -->
+        <component :is="view" v-model="model"></component>
+
+        <div class="modal-action">
+          <!-- render all actions and pass the model payload as parameter -->
+          <button
+            v-for="action in actions"
+            class="btn"
+            @click="action.callback(model)"
+          >
+            {{ action.label }}
+          </button>
+        </div>
+      </div>
+    </div>
+  </div>
+</template>
+
+<script lang="ts" setup>
+  import { reactive } from "vue";
+  import { storeToRefs } from "pinia";
+  import { useModal } from "@/stores";
+
+  const modal = useModal();
+
+  // reactive container to save the payload returned by the mounted view
+  const model = reactive({});
+
+  // convert all state properties to reactive references to be used on view
+  const { isOpen, view, actions } = storeToRefs(modal);
+</script>
+```
+
+> `<component>` is the way that `vue` handle dynamic components where `:is` receive the view that needs to be rendered, more info [here](https://vuejs.org/api/built-in-special-elements.html#component), it also support using the `v-model` directive to be able to share values dynamically, but how does this works?
+
+Once we have the view mounted inside `<component>` we can make use of `v-model` to send and update reactive references emitting events from the view we are mounting, internally from the view we want to render: `emit` a `"update:modelValue"` event that will update the reference passed via `v-model` back, eg:
+
+```html
+<!-- MyViewToRender.vue -->
+<script lang="ts" setup>
+  import { watch } from "vue";
+
+  // no need to import defineEmits
+  const emit = defineEmits(["update:modelValue"]);
+
+  // when someVar change, it will update the reference passed via v-model
+  watch(someVar, (value) => {
+    emit("update:modelValue", value);
+  });
+</script>
+```
+
+## Make use of the modal
+
+To be able to control and populate data into the modal, we'll use the store defined as follow:
+
+```html
+<template>
+  <div>
+    <button class="btn" @click="handleOnClickOpenModal">Open</button>
+  </div>
+</template>
+
+<script lang="ts" setup>
+  import useModal from "@/stores/modal";
+  import MyViewToRender from "@/views/MyViewToRender.vue";
+
+  const modal = useModal();
+  function handleOnClickOpenModal() {
+    modal.open(MyViewToRender, [
+      {
+        label: "Save",
+        callback: (dataFromView) => {
+          console.log(dataFromView);
+          modal.close();
+        },
+      },
+    ]);
+  }
+</script>
+```
+
+When a click event on the `button` occurs, it will call `handleOnClickOpenModal` which will call the action `open` from the store, which changes the `isOpen` variable from the state, and as the modal is observing this value it will render or not depending of the value as follow:
```

Comment:
```suggestion
When a click event on the `button` occurs, it will call `handleOnClickOpenModal` which at the same time will call the `open` action from the store, which changes the `isOpen` variable from the state, and as the modal is observing this value it will render or not depending of the value as follow:
```